### PR TITLE
Trade and Snapshot Data Handlers

### PR DESCRIPTION
**Pull request by Jiarui-Cao**
### Description
Data handlers responsible for writing to and reading from compressed trades/snapshots files.

### Testing
Unit tested with synthetic data and downloaded trade data for AEVOEUR.

### Impact
This should serve as the bottom-level APIs for storing/reading from compressed data files storing historical trades/snapshots data. Raw data loaders (loads trades & snapshots one-by-one) and tick (tick = 1s) data loaders (loads all trades within a tick & last snapshot within the tick) will be built on top of these handlers.

### Other
Should unblock #49 once the data loaders are built on top of this PR.

### CODE REVIEW

File: src/pysrc/data_handlers/kraken/historical/base_data_handler.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    def read_file(self, input_path: Path) -> list[TradeMessage] | list[SnapshotMessage]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def write_to_file(self, output_path: Path, data: Any) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def stream_data(
+        self, asset: str, since: datetime, until: Optional[datetime]
```
Comment (echavemann) [inline src/pysrc/data_handlers/kraken/historical/base_data_handler.py:23]: isnt `asset` an enum

File: src/pysrc/data_handlers/kraken/historical/base_data_handler.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+from pysrc.adapters.messages import TradeMessage, SnapshotMessage
+
+
+class BaseDataHandler(ABC):
+    @abstractmethod
+    def __init__(self, resource_path: Path) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def read_file(self, input_path: Path) -> list[TradeMessage] | list[SnapshotMessage]:
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/base_data_handler.py:14]: maybe type to `list[any]`? or alternatively type `data` in `write_to_file` to `list[TradeMessage] | list[SnapshotMessage]`

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+class SnapshotsDataHandler(BaseDataHandler):
+    def __init__(self, resource_path: Path) -> None:
+        self.resource_path = resource_path
+        self._zstd_options = {CParameter.compressionLevel: 10}
+        self._metadata_size = 24
+
+    def _compress(self, data: bytes, output_path: Path) -> None:
+        with open(output_path, "wb") as f:
+            f.write(compress(data, level_or_option=self._zstd_options))
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:21]: use `ZstdCompressor` (https://pyzstd.readthedocs.io/en/stable/#streaming-compression), you get way better compression ratio

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+    def _compress(self, data: bytes, output_path: Path) -> None:
+        with open(output_path, "wb") as f:
+            f.write(compress(data, level_or_option=self._zstd_options))
+
+    def _decompress(self, input_path: Path) -> bytes:
+        with open(input_path, "rb") as f:
+            return decompress(f.read())
+
+    def _serialize_csv(self, csv_path: Path) -> Path:
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:29]: delete, not used

**Comment by echavemann** [comment]
As I understand these are the lower level ones? @jerry

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                time, asset, 1, price, volume, OrderSide.BID, Market.KRAKEN_SPOT
+            )
+            trades.append(trade)
+        return trades
+
+    def write_to_file(self, output_path: Path, data: list[TradeMessage]) -> None:
+        arr = self._trades_to_numpy(data)
+        self._compress(arr.tobytes(), output_path)
+
+    def _decompress_bytes(
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:57]: based on the snapshot handler i thought ZstdFile would do this for you?

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        if not raw_data:
+            return None
+        elif len(raw_data) < self._record_size:
+            raise ValueError("Failed to read data from stream")
+
+        time = struct.unpack("Q", raw_data[0:8])[0]
+        price = struct.unpack("f", raw_data[8:12])[0]
+        volume = struct.unpack("f", raw_data[12:16])[0]
+
+        return TradeMessage(
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:84]: I think this is specific to the trades downloaded from kraken, but would be nice if we generalized this to not assume BID and KRAKEN_SPOT. also, I we should be consistent with snapshot, so either turn this into a to_bytes() method in TradeMessage, or move the to_bytes() method for snapshot into the data handler

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        decompressor = EndlessZstdDecompressor()
+        for cur_path in file_paths:
+            with open(cur_path, "rb") as f:
+                while True:
+                    trade = self._trade_message_from_stream(decompressor, f, asset)
+                    if not trade:
+                        break
+                    yield trade
+
+    def _trade_message_from_stream_2(
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:116]: what is this

**Comment by Jiarui-Cao** [comment]
Um actually read & write are lower level ones, stream is higher level. I was thinking stream needs to iterate across multiple days. Or do we want that to be handled by the data loaders above & have these handlers only iterate 1 file?

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py (hunk 7 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:116]: my bad some historical issues, will fix

Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:116]: I was testing the speed of 2 implementations, I think stream_data_2 is faster

File: src/pysrc/data_handlers/kraken/historical/base_data_handler.py (hunk 2 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/base_data_handler.py:14]: I'll do the former, the latter gives "This violates the Liskov substitution principle" warning

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py (hunk 4 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:29]: noted

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py (hunk 5 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:57]: Ah historical issue, I'll remove this and `stream_data` and replace with the implementation in `stream_data_2`. My bad for forgetting to clean this

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 3 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:21]: ok

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py (hunk 6 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:84]: I can bake to_bytes() into TradeMessage. Also can add side & market fields to compressed data

**Comment by Jiarui-Cao** [comment]
For trades, I added a "side" field to compressed data (so 4 field in total: time, price, quantity, side), but didn't include a market field. I assumed that all trades within a same file should be from the same asset and market, so storing market within the .bin files would be redundant, but correct me if I'm wrong

**Comment by rewong03** [comment]
> For trades, I added a "side" field to compressed data (so 4 field in total: time, price, quantity, side), but didn't include a market field. I assumed that all trades within a same file should be from the same asset and market, so storing market within the .bin files would be redundant, but correct me if I'm wrong

hmm i think realistically that's the case but idk if we want this layer to make that assumption. hopefully if we do include it it should get compressed away, but i think at the very least you should include the market somewhere in the file instead of making it default to `SPOT` like you currently have. thoughts @echavemann ?

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+from typing import Optional, Generator
+
+from pysrc.data_handlers.kraken.historical.base_data_handler import BaseDataHandler
+from pysrc.adapters.messages import SnapshotMessage
+from pysrc.util.types import Market
+
+
+class SnapshotsDataHandler(BaseDataHandler):
+    def __init__(self, resource_path: Path, write_size: int = 1000) -> None:
+        self.resource_path = resource_path
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:14]: add underscores in front of class variable names

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        count = 0
+        with open(output_path, "wb") as f:
+            for snapshot in data:
+                out += snapshot.to_bytes()
+                count += 1
+                if count >= self.write_size:
+                    f.write(compressor.compress(out, ZstdCompressor.FLUSH_FRAME))
+                    out = b""
+                    count = 0
+            f.write(compressor.compress(out, ZstdCompressor.FLUSH_FRAME))
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:41]: it looks like you're manually buffering writes, was there any particular reason why?

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        return SnapshotMessage(
+            time=time,
+            feedcode=feedcode_data.decode(),
+            market=Market(market_value),
+            bids=bids.reshape((-1, 2)).tolist(),
+            asks=asks.reshape((-1, 2)).tolist(),
+        )
+
+    def stream_read(self, input_path: Path) -> Generator[SnapshotMessage, None, None]:
+        if not input_path.exists():
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:79]: i would either add this check to `read` as well or just remove

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        arr = np.empty(len(trades), dtype=self._np_dtype)
+        for i, trade in enumerate(trades):
+            arr[i] = (trade.time, trade.price, trade.quantity, trade.side.value)
+        return arr
+
+    def read(self, input_path: Path) -> list[TradeMessage]:
+        if not input_path.exists():
+            raise ValueError(f"Expected file '{input_path}' does not exist")
+        trades = []
+        asset = kraken_to_asset(input_path.parent.name)
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:36]: this is making assumptions about the directory structure, i don't think we want to make those assumptions at this layer of abstraction

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        bids = np.frombuffer(bids_bytes)
+
+        asks_bytes = file.read(asks_size)
+        if len(asks_bytes) < asks_size:
+            raise ValueError("Failed to read asks from stream")
+        asks = np.frombuffer(asks_bytes)
+
+        return SnapshotMessage(
+            time=time,
+            feedcode=feedcode_data.decode(),
```
Comment (rewong03) [inline src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:72]: just as general note we will need to change this to an asset in the future once we have the asset -> future name mappings

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 9 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:41]: testing with 3,000,000 trades without this buffer, somehow it just spends forever to run, especially when the `out` binary string grows to be too large. I actually don't know the reason why but buffering seems to fix that

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 10 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:79]: I thought we want a streaming functionality in this class as opposed to reading the entire file?

File: src/pysrc/data_handlers/kraken/historical/trades_data_handler.py (hunk 11 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/trades_data_handler.py:36]: I think the solution would be to either make `asset` an argument to `read()` or store `asset` in the compressed file. That raises a similar issue with storing `market`. I think it's better to make them arguments to the function as opposed to storing them in the compressed file. Thoughts? @echavemann @rewong03

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 12 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:72]: cool

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 8 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:14]: got

Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:14]: it

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 9 above)
Comment (rewong03) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:41]: what's the performance if you don't buffer but also don't pass FLUSH_FRAME?

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 10 above)
Comment (rewong03) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:79]: oh yeah i meant that read should also either check that the file exists before trying to read, or both functions shouldn't check

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 9 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:41]: the performance degradation didn't come from compressor.compress, but instead `out += trade.to_bytes()`

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 10 above)
Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:79]: oh crap I forgot to check in `read`

File: src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py (hunk 9 above)
Comment (rewong03) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:41]: why do we keep increasing the size of out? why not something like

```
for snapshot in data:
  out = snapshot.to_bytes()
  write out to file
```

Comment (Jiarui-Cao) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:41]: like I think with 100,000 trades it took less than 1s, but once I bump to 500,000 trades it takes more than 10s

Comment (rewong03) [reply src/pysrc/data_handlers/kraken/historical/snapshots_data_handler.py:41]:
```
compressor = ZstdCompressor(level_or_option=self._zstd_options)
out = b""
with open(output_path, "wb") as f:

    for trade in data:
        out = trade.to_bytes()
        f.write(compressor.compress(out))

    f.write(compressor.flush())
```

with that i was able to get ~3x improvement

**Pull request closed without merging**